lib/reconcile/utils.go

```diff
@@ -580,3 +661,55 @@ func getBaseAstarteAPIURL(cr *apiv1alpha1.Astarte) string {
 
 	return fmt.Sprintf("%s://%s", scheme, cr.Spec.API.Host)
 }
+
+func handleGenericUserCredentialsSecret(username, password, usernameKey, passwordKey, secretName string, cr *apiv1alpha1.Astarte, c client.Client, scheme *runtime.Scheme) error {
+	createUserCredentialsSecret := true
+	createUserCredentialsSecretFromCredentials := false
+	if cr.Spec.Cassandra.Connection != nil {
+		if cr.Spec.Cassandra.Connection.Secret != nil {
+			createUserCredentialsSecret = false
+		} else if username != "" {
+			createUserCredentialsSecretFromCredentials = true
+		}
+	}
+
+	if createUserCredentialsSecret {
```

Comment:
There is one caveat here. Depending on the situation, the secret should or shouldn't be created at all costs. So you should pass to this function an additional `bool` that, depending on the value, will either force the secret to be created with a random password or not be created at all.

As such, you should check if the user specified a Username/Password combination first (which you can do). If he didn't AND the bool is false, then you should not create anything.